### PR TITLE
:clock3: Add reception and surgery opening times

### PR DESCRIPTION
**Pull request by markysoft**
Moved GP Count logic from renderer to middleware.
Exposed opening times on profiles page.

### CODE REVIEW

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:warning: deployment of 52 for nhsuk/profiles failed

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

File: app/lib/parseOpeningTimes.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  if (!allTimes) {
+    return undefined;
+  }
+  return {
+    reception: allTimes.reception && parseWeek(allTimes.reception),
+    surgery: allTimes.surgery && parseWeek(allTimes.surgery),
+  };
+}
+
+module.exports = { parseAll, parseWeek, parseDay };
```
Comment (st3v3nhunt) [inline app/lib/parseOpeningTimes.js]: I'd prefer this spread over multiple lines as future changes are easy to see via diffs e.g.
```
module.exports = {
  parseAll,
  parseWeek,
  parseDay,
};
````

Do `parseWeek` and `parseDay` need to be exported?

File: app/lib/timeUtils.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  } else if (hours === 0) {
+    hours = 12;
+  }
+
+  const mins = hoursMins[1] === '00' ? '' : `:${hoursMins[1]}`;
+
+  return `${hours}${mins}${suffix}`;
+}
+
+module.exports = { toAmPm };
```
Comment (st3v3nhunt) [inline app/lib/timeUtils.js]: I'd prefer this spread over multiple lines as future changes are easy to see via diffs

File: app/views/includes/opening-times.nunjucks [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  </table>
+{% endmacro %}
+
+{% if gp.openingTimes %}
+  {{ openingTimes(gp.openingTimes.reception, 'When reception is open') }}
+  {{ openingTimes(gp.openingTimes.surgery, 'When GPs see patients') }}
+{% else %}
+  {{ openingTimesUnavailable('When reception is open', 'opening times') }}
+  {{ openingTimesUnavailable('When GPs see patients', 'when you can get a GP appointment') }}
+{% endif %}
```
Comment (st3v3nhunt) [inline app/views/includes/opening-times.nunjucks:31]: Newline please

File: test/unit/lib/parseOpeningTimes.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    it('should change empty array to closed message', () => {
+      const openingTimes = parseOpeningTimes.parseDay([]);
+      // eslint-disable-next-line no-unused-expressions
+      expect(openingTimes.length).to.equal(1);
+      expect(openingTimes[0]).to.equal('Closed');
+    });
+  });
+});
+
```
Comment (st3v3nhunt) [inline test/unit/lib/parseOpeningTimes.js]: Remove additional line please

File: test/unit/lib/parseOpeningTimes.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+const expect = chai.expect;
+
+describe('parseOpeningTimes', () => {
+  describe('parseAll', () => {
+    it('should return undefined for empty opening times', () => {
+      const openingTimes = parseOpeningTimes.parseAll(undefined);
+      // eslint-disable-next-line no-unused-expressions
+      expect(openingTimes).to.be.undefined;
+    });
+    it('should gracefully handle missing days on opening times', () => {
```
Comment (st3v3nhunt) [inline test/unit/lib/parseOpeningTimes.js]: Additional line please

File: test/unit/lib/timeUtils.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      expect(time).to.equal('12pm');
+    });
+
+    it('it should return invalid strings unchanged', () => {
+      const time = timeUtils.toAmPm('notATime');
+      expect(time).to.equal('notATime');
+    });
+  });
+});
+
```
Comment (st3v3nhunt) [inline test/unit/lib/timeUtils.js]: Remove additional line please

File: app/lib/parseOpeningTimes.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  daysOfWeek.forEach((day) => {
+    const daySessions = times[day.toLowerCase()];
+    const sessions = parseDay(daySessions);
+    parsedTimes.push({ day, sessions });
+  });
+  return addPadding(parsedTimes);
+}
+
+function parseAll(allTimes) {
+  if (!allTimes) {
```
Comment (st3v3nhunt) [inline app/lib/parseOpeningTimes.js]: Use ternary to have a single return statement?

File: app/lib/parseOpeningTimes.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    return undefined;
+  }
+  return {
+    reception: allTimes.reception && parseWeek(allTimes.reception),
```
Comment (st3v3nhunt) [inline app/lib/parseOpeningTimes.js]: I can't see a test for the null check for either `allTimes.reception` or `allTimes.surgery` (on the following line)

File: test/unit/lib/parseOpeningTimes.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        saturday: twoSessionDay,
+        sunday: twoSessionDay,
+      };
+
+      const days = parseOpeningTimes.parseWeek(week);
+      expect(days.length).to.equal(7);
+      // eslint-disable-next-line no-unused-expressions
+      expect(days[0].padding).to.be.undefined;
+      expect(days[3].padding).to.equal(1);
+      // expect(openingTimes[1]).to.equal('1:30pm to 8:00pm');
```
Comment (st3v3nhunt) [inline test/unit/lib/parseOpeningTimes.js]: Should this be deleted?

File: test/unit/lib/parseOpeningTimes.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    });
+
+    it('should replace contiguous times with a single message', () => {
```
Comment (st3v3nhunt) [inline test/unit/lib/parseOpeningTimes.js]: Looks like a duplicate of https://github.com/nhsuk/profiles/pull/52/files#diff-19ef8052a34c4516c393f39bb38bb599R119

File: test/unit/lib/parseOpeningTimes.js [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        }
+      ];
+
+      const openingTimes = parseOpeningTimes.parseDay(monday);
+      // eslint-disable-next-line no-unused-expressions
+      expect(openingTimes.length).to.equal(1);
+      expect(openingTimes[0]).to.equal('7:30am to 8pm');
+    });
+
+    it('should change empty array to closed message', () => {
```
Comment (st3v3nhunt) [inline test/unit/lib/parseOpeningTimes.js:118]: Looks like a duplicate of https://github.com/nhsuk/profiles/pull/52/files#diff-19ef8052a34c4516c393f39bb38bb599R137

File: test/integration/openingTimes.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+describe('app', () => {
+  describe('opening times', () => {
+    it('should return reception and surgery opening times', (done) => {
+      chai.request(app)
+        .get(`${constants.SITE_ROOT}/43213`)
+        .end((err, res) => {
+          expect(err).to.equal(null);
+          expect(res).to.have.status(200);
+
+          expect(res.text).to.include('When reception is open');
```
Comment (st3v3nhunt) [inline test/integration/openingTimes.js:37]: Can cheerio be used to check this text and be a bit more specific about the locator used to access it? Same goes for the line below.

File: test/integration/openingTimes.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    it('should return call reception message for no reception or surgery opening times', (done) => {
+      chai.request(app)
+        .get(`${constants.SITE_ROOT}/107891`)
+        .end((err, res) => {
+          expect(err).to.equal(null);
+          expect(res).to.have.status(200);
+
+          expect(res.text).to.include('When reception is open');
```
Comment (st3v3nhunt) [inline test/integration/openingTimes.js:37]: Can cheerio be used to check this text and be a bit more specific about the locator used to access it? Same goes for the line below.

File: test/unit/lib/timeUtils.js (hunk 6 above)
Comment (st3v3nhunt) [inline test/unit/lib/timeUtils.js]: Remove this line please

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
I'm not a fan of the name `gpCountsHelper`, I'm not sure I would see that file and go
> aha, that will have a method in to work out if there are any GPs available

Suggest renaming it to `gpHelper` or something similar. I'd also be inclined to change the name of the function `gpsAvailable` to `areGpsAvailable`

I like the lib for converting times to 12 hour mode. Probably something to think about sharing in the future.

File: app/lib/parseOpeningTimes.js (hunk 1 above)
Comment (markysoft) [reply app/lib/parseOpeningTimes.js]: parseWeek and parseDay are exported to test individually

File: app/lib/parseOpeningTimes.js (hunk 8 above)
Comment (markysoft) [reply app/lib/parseOpeningTimes.js]: added check

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployed to [http://profiles-pr-52.dev.c2s.nhschoices.net](http://profiles-pr-52.dev.c2s.nhschoices.net)

**Review by st3v3nhunt — APPROVED** [review]

**Pull request merged**